README.md
# fgj-15
Finnish Game Jam 2015 game

# Trello link https://trello.com/b/OfWNnufS/fgj2015
# Pelattavissa http://users.jyu.fi/~sakamiai/fgj/lifecycle-manager/


JoyToKey_en/Changes.txt
---------------------------
  JoyToKey change history  
---------------------------

Ver 5.6
* Enhanced "Keyboard 2" functionality to support the followings
  - support 4 keys per definition
  - support the rotation of 4 input definitions
  - support auto repeat and toggle configuration

* Upgraded DirectX library version from 7 to 8.


Ver 5.5.x
* bug fixes


Ver 5.5
* Support "Button Number Mapping"

  Example Usage1:
    Swap button numbers if the connected joystick has a different layout
    from your favorite one.

  Example Usage2:
    Associate Button21 with Button1 so that both functions are triggered
    when Button1 is pressed.

  Example Usage3:
    It is also possible to configure some button (e.g. Button22) to be
    triggered when both Button5 and Button6 are pressed simultaneously.

* If some buttons are pressed when profile is changed or SHIFT function
  is triggered, those button press are likely to trigger both old and new
  assignments right before/after the switch.

  It is now possible to define certain duration (e.g. 100 msec) to ignore
  those inputs right after the switch (from preference menu)

* When both mouse cursor movement and mouse click are assigned to a button,
  click&hold will be processed before cursor movement.

* Improved the reliability after resuming from PC sleep.

* Preferences were moved from menu items to a separate window.

* Added an ability to specify the number of buttons to configure.
  (by default 16 buttons, MAX 32 buttons)

* Changed the user interface font to Calibri

* Configuration files are now saved in Unicode format (UTF8+BOM).
  This may fix garbled character issues for the use of non-default
  language.


Ver 5.4.2
* Fixed garbled characters for some non-Japanese PC and keyboard.

* (only for advanced users)
  Support specifying "JoyToKey.ini" file path via registry.
  If you have any reason why you need to forcibly specify "JoyToKey.ini"
  file path (e.g. you're using some launcher like Hyperlaunch), you can
  do so via registry editor.
  1. Run "regedit" program
  2. Create a key "HKEY_CURRENT_USER\Software\JoyToKey"
  3. Create a string value "IniFilePath=c:\whateverpath\JoyToKey.ini"


Ver 5.4.1
* Support various file formats (such as mp3) for sound notifications.


Ver 5.4
* Added an ability to play sound or show tasktray balloon when the
  selected profile is changed.

  - Tasktray balloon can be enabled from "Preferences" menu.

  - Sound notification can be configured for each profile in "Options"
    tab. (currently, only WAV files are supported)

  For example, sound notification could be useful for visually impaired
  users, while they step through different profiles to help them to pick
  the right one.

* The size of profile files is made smaller by eliminating empty lines.

* In case "JoyToKey.ini" file or its directory is not write-able,
  JoyToKey will try to make them write-able.


Ver 5.3.1
* Homepage URL was changed to http://joytokey.net/


Ver 5.3.0
* Support delayed input for a special function to switch profiles.
  If you want to keep JoyToKey running while you're playing a game
  which has a native joystick support, the followings may be useful.
  - Temporarily switch to a profile, which is basically empty
  - But it has one button assigned for switching back to the
    original profile after being pressed for X seconds.

* Control the behaviour when JoyToKey window is active from menu
  "Preferences" ->
    "Don't emulate keyboard input when JoyToKey window is active"
  (by default true)


Ver 5.2.3
* Support emulation for 4th and 5th mouse button.
  They could be useful if you're using mouse gestures with joysticks.

* Turbo mode was enhanced to support up to x16 processing speed.
  However, this is experimental and should not be used for regular use
  as it's very CPU intensive and is likely to slow down the PC.


Ver 5.2.2
* Support for loading configuration files from a common AppData folder
  when it exists. (e.g. C:\ProgramData\JoyToKey)

* Added a new menu "Open file folder in Explorer"


Ver 5.2.1
* Bug fix: garbled characters for allow keys


Ver 5.2.0
* Support the combination of keyboard and mouse click
  Ex) SHIFT+Click, CTRL+Click

  In order to specify a mouse click in keyboard assignment window,
  please right-click the entry box.

* Bug fix: minor bug fix for 5.1.2


Ver 5.1.2
* Bug fix: SHIFT + arrow keys were not properly processed.


Ver 5.1.1
* Mouse cursor emulation with analog sticks is further improved.


Ver 5.1.0
* Mouse cursor emulation is now much smoother with analog sticks,
  by using a continuous acceleration without a threshold.

* In case analog stick input has been wrong, axis mapping can be
  manually specified.
  (Menu: "Preferences" -> "Configure joysticks")

* In case you have multiple joysticks and want to always use one
  of them as "Joystick 2" (or any other joystick number), you can
  define a priority/preferred number for each device.
  (Menu: "Preferences" -> "Configure joysticks")

* Miscellaneous bug fixes.


Ver 5.0.3
* (Improvement) Re-detect joysticks after resuming from sleep.


Ver 5.0.2
* Bug fix: Preference "Highlight buttons when JoyToKey is active" was not
  properly saved.


Ver 5.0.1
* Added a menu "Help -> Overview"
* Added a button "Edit button assignment"
* Minor bug fixes


Ver 5.0
* Support for different key assignments based on analog stick input level.

* Support for different key assignments based on button hold duration.

* Support for rotating key assignments.  For example, whenever the same
  button is pressed, its key assignment can rotate like 
  "A -> B -> A -> B -> A ..."

* Enhanced auto-repeat key functionality.
  The following auto-repeat definitions are newly possible:
  - "1.5 repeats per second"
  - "one repeat per 123.4 seconds"
  - "auto-repeat only after 500 milli-seconds"

* When the main window of JoyToKey is activated and joystick buttons are
  pressed, those buttons will be highlighted in the main window so that

phaser-fay/PurpleBall.hs
{-# LANGUAGE RecordWildCards #-}
module PurpleBall where

import Prelude
import Data.Function
import DOM
import PhaserHS
import Arrows
import Utils

playerSpeed = 180

--assetDir = ""
assetDir = "./../../assets/"

data Player = Player
    { sprite     :: Sprite
    , texPrefix  :: String
    , swordTimer :: Int
    , hurtTimer  :: Int
    }

data GameState = GameState
    { physics   :: Physics
    , player1   :: Player
    , player2   :: Player
    , platforms :: Group
    , ball      :: Sprite
    }


-- | Preloads assets.
preloadGame :: Game -> Fay ()
preloadGame game = do
    putStrLn "Preloading....."
    mapM_ (uncurry $ loadImage game)
        [("ground", assetDir ++ "platform.png")
        ,("star", assetDir ++ "star.png")
        ,("bluepart", assetDir ++ "bluepart.png")
        ,("redpart", assetDir ++ "redpart.png")]
    loadSpriteSheet game "bluedude" (assetDir ++ "blueplayer.png") (20, 32)
    loadSpriteSheet game "reddude" (assetDir ++ "redplayer.png") (20, 32)
    loadSpriteSheet game "bluesword" (assetDir ++ "bluesword.png") (40, 32)
    loadSpriteSheet game "redsword" (assetDir ++ "redsword.png") (40, 32)
    loadSpriteSheet game "ball" (assetDir ++ "ballz.png") (16, 32)
    loadBitmapFont game "testfont" (assetDir ++ "font.png") (assetDir ++ "font.fnt")


changeTexture :: Player -> String -> String -> Fay ()
changeTexture player texSuffix animName = do
    loadTexture (sprite player) (texPrefix player ++ texSuffix)
    createAnim animName
    playAnimation (sprite player) animName
    where
        createAnim :: String -> Fay ()
        createAnim "walk" = addAnimation (sprite player) "walk" [0 .. 7] 10 True
        createAnim "hit" = addAnimation (sprite player) "hit" ([0 .. 4] ++ [4, 4, 4, 4] ) 10 True


-- | Creates player sprite.
createPlayer :: Game
             -> Physics
             -> (Double, Double) -- ^ Player start position.
             -> (Double, Double) -- ^ Player score text position.
             -> String -- ^ Player texture name prefix.
             -> Fay Player
createPlayer game physics startPos textPos texPre = do
    sprite <- newSprite game (texPre ++ "dude") (50, 50)
    enable physics sprite
    sprite ~> (body >>> gravity >>> setY 600)
    sprite ~> (anchor >>> setTo (0.5, 0.5))
    sprite ~> (position >>> setTo startPos)
    sprite ~> (scale >>> setTo (3, 3))
    sprite ~> (body >>> flip collideWorldBounds True)

    return $ Player sprite texPre 0 0


-- | Creates initial game state.
createGame :: Game -> Fay GameState
createGame game = do
    putStrLn "Creating..."
    startGamePad game        -- Enables gamepad support.
    setSmoothing game False  -- Enables crisp pixel graphics.
    physics <- startPhysics game arcadePhysics

    platforms <- newGroup game
    enableBody platforms True

    -- Create platforms.
    let makep = platf platforms
    makep (400, 600) (800, 32)

    -- Middle plaftorms.
    makep (400, 116) (64, 16)
    makep (400, 300) (64, 16)
    makep (400, 480) (64, 16)

    -- Side platforms
    forM_ [230, 400] $ \i -> do
        makep (100, i) (200, 16)
        makep (700, i) (200, 16)

    -- Create players.
    let cp = createPlayer game physics
    player1' <- cp (740, 500) (710, 50) "blue"
    player2' <- cp (60, 500) (30, 50) "red"

    -- Set default animations.
    forM_ [player1', player2'] $ \p -> changeTexture p "dude" "walk"

    ball <- newSprite game "ball" (400, 60)
    ball ~> (anchor >>> setTo (0.5, 0.5))
    ball ~> (scale >>> setTo (3, 3))
    addAnimation ball "bounce" [0 .. 12] 15 True
    playAnimation ball "bounce"

    txt <- newText game "testfont" 64 (200, 100) "Run to the\npurple ball!"
    singleShot game (3 * seconds) $ destroy txt

    return $ GameState physics player1' player2' platforms ball
    where
        createStar :: Game -> Group -> Fay ()
        createStar game stars = do
            rBounce <- fmap (* 0.2) random -- Random bounciness.
            rx      <- fmap (\i -> 80 + i * (800 - 80 * 2)) random -- Random x position.
            star <- create stars "star" (rx, 0)
            star ~> (anchor >>> setTo (0.5, 0.5))
            star ~> (scale >>> setTo (3, 3))
            star ~> (body >>> gravity >>> setY 80)
            star ~> (body >>> bounce >>> setY (0.7 + rBounce))

        platf :: Group -> (Double, Double) -> (Double, Double) -> Fay()
        platf gr pos (w, h) = do
            ground <- create gr "ground" (0, 0)
            ground ~> (anchor >>> setTo (0.5, 0.5))
            ground `setWidth` w
            ground `setHeight` h
            ground ~> (position >>> setTo pos)
            setImmovable (body ground) True


type PlayerUpdater = State GameState -> (Player -> Player) -> Fay ()


-- | Updates player 1 state.
updatePlayer1 :: PlayerUpdater
updatePlayer1 state updateFunc =
    modify state $ \g -> g { player1 = updateFunc $ player1 g }


-- | Updates player 2 state.
updatePlayer2 :: PlayerUpdater
updatePlayer2 state updateFunc =
    modify state $ \g -> g { player2 = updateFunc $ player2 g }


-- | Updates game state.
updateGame :: Game -> State GameState -> Fay ()
updateGame game state = do
    GameState{..} <- get state
    let collider   = collide physics

    -- Physics.
    collider (sprite player1) platforms
    collider (sprite player2) platforms

    swordFight player1 player2 updatePlayer2
    swordFight player2 player1 updatePlayer1

    checkEnd ball player1 1
    checkEnd ball player2 2

    -- Update both players.
    forM_ (enumerate [(player1, updatePlayer1), (player2, updatePlayer2)]) $ \(i, (player, updater)) -> do
        input <- getGamePadInput game i

        -- Player is hurt, attacking or moving.

        -- Hurt update.
        when (hurtTimer player > 0) $
            updater state $ \p -> p { hurtTimer = hurtTimer p - 1 }

        -- Sword update.
        when (swordTimer player > 0 && hurtTimer player == 0) $ do
            updater state $ \p -> p { swordTimer = swordTimer p - 1 }
            when (swordTimer player - 1 == 0) $ changeTexture player "dude" "walk"
            (sprite player) ~> (body >>> velocity >>> setX 0)

        -- Regular update.
        when (swordTimer player == 0 && hurtTimer player == 0) $
            updatePlayer input player updater

        -- Jumping.
        let onGround = player ~> (sprite >>> body >>> touchingDown)
        when (padUp input && onGround) $
            player ~> (sprite >>> body >>> velocity >>> setY (-400))

    where
        enumerate :: [a] -> [(Int, a)]
        enumerate = zip [1..]

        -- Ends game when player is outside of screen.
        checkEnd :: Sprite -> Player -> Int -> Fay ()
        checkEnd ball Player{..} index = do
            let (ax, ay)    = sprite ~> (position >>> vectorToTuple)
                (dx, dy)    = ball ~> (position >>> vectorToTuple)
                distSq      = (dx - ax) ** 2 + (dy - ay) ** 2
                closeEnough = distSq < 30 ** 2
            when closeEnough $ do
                destroy game
                endGame index

        swordFight :: Player -> Player -> PlayerUpdater -> Fay ()
        swordFight attacker defender defenderUpdater =
            let (ax, ay)    = attacker ~> (sprite >>> position >>> vectorToTuple)
                (dx, dy)    = defender ~> (sprite >>> position >>> vectorToTuple)
                distSq      = (dx - ax) ** 2 + (dy - ay) ** 2
                closeEnough = distSq < 70 ** 2
                attacking   = swordTimer attacker > 20
                force = sign (dx - ax) * 300
            in when (closeEnough && attacking) $ do
                (sprite defender) ~> (body >>> velocity >>> setX force)
                defenderUpdater state $ \p -> p { hurtTimer = 20 }

        updatePlayer :: GamePadInput -> Player -> PlayerUpdater -> Fay ()
        updatePlayer pad player@Player{..} update
            | pad ~> padA || pad ~> padB = do
                changeTexture player "sword" "hit"
                update state $ \p -> p { swordTimer = 40 }
            | pad ~> padRight = do
                sprite ~> (body >>> velocity >>> setX playerSpeed)
                sprite ~> (scale >>> setX 3)
                sprite `playAnimation` "walk"
            | pad ~> padLeft = do
                sprite ~> (body >>> velocity >>> setX (-playerSpeed))
                sprite `playAnimation` "walk"
                sprite ~> (scale >>> setX (-3))
            | otherwise = do
                when (swordTimer == 0) $
                    sprite `setFrame` 2
                sprite ~> (body >>> velocity >>> setX 0)


-- | Initializes Phaser and starts game.
startGame :: Event -> Fay ()
startGame event = do
    game <- newGame 800 600
    newState game "Mainmenu" preloadGame createGame updateGame
    changeState game "Mainmenu"
    return ()


main :: Fay ()
main = addWindowEvent "load" startGame
